### ISSUE TITLE
Poor performance after fine-tuning with 50 epochs on 20-class human parsing dataset

### ISSUE THREAD

**Issue by ShulingTang** [facebookresearch/sapiens#140]
Description: Hello,
I have a dataset containing 50,000 samples for 20-class human parsing. I followed the [Body Part Segmentation](https://github.com/facebookresearch/sapiens/blob/main/docs/finetune/SEG_README.md) guide to fine-tune the 1b model. After training for 50 epochs, the results are still unsatisfactory.

Here are the modifications I made:

File: /seg/configs/sapiens_seg/seg_face/sapiens_1b_seg_face-1024x768.py

- num_epochs
- pretrained_checkpoint
- num_classes
- RandomFlip
- data_root

File: /seg/mmseg/datasets/face.py
- CLASSES

File: seg/scripts/finetune/seg_face/sapiens_1b/node.sh
- LOAD_FROM
- OUTPUT_DIR
Here are some examples of the training results after 50 epochs:

![000001](https://github.com/user-attachments/assets/452ffd40-1708-477c-b86b-76a841bcf0bb)

![000050](https://github.com/user-attachments/assets/ddc4d184-2cf9-4d7d-9be8-b77072b1fc21)

![0000099]https://images.shuling.top/2024/10/bae5f6a12bb5943b07d746c2b7625a51.jpg)

![](https://images.shuling.top/2024/10/f13706d87425039ac2ea8eac6f1aa3d7.png)

I would like to ask if there is any issue with my settings, or why the results did not meet my expectations?
Thank you for your help!

**Comment by rawalkhirodkar** [comment]
@ShulingTang Hello, can you share the visualizations on the train samples? You should find them under $OUTPUT_DIR/vis_data. On the train data, do the results make sense?

**Comment by ShulingTang** [comment]
> @ShulingTang Hello, can you share the visualizations on the train samples? You should find them under $OUTPUT_DIR/vis_data. On the train data, do the results make sense?

Hello,
I used $SAPIENS_ROOT/seg/scripts/demo/local/seg.sh to load the checkpoint from my training to evaluate the results. After testing the model trained for 77 epochs, the results are still the same as those from 50 epochs.

However, when I checked the results under $OUTPUT_DIR/vis_data, they seemed to show good training performance. Below are some example images from the visualized data:

![train_103100](https://github.com/user-attachments/assets/da0c2aed-6897-4cce-a1e4-ed77bacc538d)

![train_801500](https://github.com/user-attachments/assets/338e918c-d853-4d29-8622-c70b006d4e11)

Am I using the inference script incorrectly?
The model I loaded is the latest epoch_xx.pth from $OUTPUT_DIR.
Thank you for your assistance!

**Comment by rawalkhirodkar** [comment]
@ShulingTang The train results look good, likely the inference has issue.

Can you share your seg.sh? You need to change both config file and model checkpoint.

**Comment by ShulingTang** [comment]
The following is my seg.sh file and config file. Are there any wrong settings?

seg.sh:
```
#!/bin/bash

cd ../../.. || exit
SAPIENS_CHECKPOINT_ROOT=/home/${USER}/sapiens_host

#----------------------------set your input and output directories----------------------------------------------
INPUT='../pose/demo/data/itw_videos/reel1'
#OUTPUT="/home/${USER}/Desktop/sapiens/seg/Outputs/vis/itw_videos/reel1_seg"
OUTPUT="/path/epoch_77"
#--------------------------MODEL CARD---------------
# MODEL_NAME='sapiens_0.3b'; CHECKPOINT=$SAPIENS_CHECKPOINT_ROOT/seg/checkpoints/sapiens_0.3b/sapiens_0.3b_goliath_best_goliath_mIoU_7673_epoch_194.pth
# MODEL_NAME='sapiens_0.6b'; CHECKPOINT=$SAPIENS_CHECKPOINT_ROOT/seg/checkpoints/sapiens_0.6b/sapiens_0.6b_goliath_best_goliath_mIoU_7777_epoch_178.pth
#MODEL_NAME='sapiens_1b'; CHECKPOINT=$SAPIENS_CHECKPOINT_ROOT/seg/checkpoints/sapiens_1b/sapiens_1b_goliath_best_goliath_mIoU_7994_epoch_151.pth
MODEL_NAME='sapiens_1b'; CHECKPOINT=/path/model_checkpoint/sapiens/epoch_77.pth
# MODEL_NAME='sapiens_2b'; CHECKPOINT=$SAPIENS_CHECKPOINT_ROOT/seg/checkpoints/sapiens_2b/sapiens_2b_goliath_best_goliath_mIoU_8131_epoch_200.pth

DATASET='goliath'
MODEL="${MODEL_NAME}_${DATASET}-1024x768"
CONFIG_FILE="configs/sapiens_seg/${DATASET}/${MODEL}.py"
OUTPUT=$OUTPUT/$MODEL_NAME

##-------------------------------------inference-------------------------------------
RUN_FILE='demo/demo_seg_vis.py'

## number of inference jobs per gpu, total number of gpus and gpu ids
JOBS_PER_GPU=1; TOTAL_GPUS=1; VALID_GPU_IDS=(0 1 2 3 4 5 6 7)
TOTAL_JOBS=$((JOBS_PER_GPU * TOTAL_GPUS))

# Find all images and sort them, then write to a temporary text file
IMAGE_LIST="${INPUT}/image_list.txt"
find "${INPUT}" -type f \( -iname \*.jpg -o -iname \*.png \) | sort > "${IMAGE_LIST}"

# Check if image list was created successfully
if [ ! -s "${IMAGE_LIST}" ]; then
  echo "No images found. Check your input directory and permissions."
  exit 1
fi

# Count images and calculate the number of images per text file
NUM_IMAGES=$(wc -l < "${IMAGE_LIST}")
IMAGES_PER_FILE=$((NUM_IMAGES / TOTAL_JOBS))
EXTRA_IMAGES=$((NUM_IMAGES % TOTAL_JOBS))

export TF_CPP_MIN_LOG_LEVEL=2
echo "Distributing ${NUM_IMAGES} image paths into ${TOTAL_JOBS} jobs."

# Divide image paths into text files for each job
for ((i=0; i<TOTAL_JOBS; i++)); do
  TEXT_FILE="${INPUT}/image_paths_$((i+1)).txt"
  if [ $i -eq $((TOTAL_JOBS - 1)) ]; then
    # For the last text file, write all remaining image paths
    tail -n +$((IMAGES_PER_FILE * i + 1)) "${IMAGE_LIST}" > "${TEXT_FILE}"
  else
    # Write the exact number of image paths per text file
    head -n $((IMAGES_PER_FILE * (i + 1))) "${IMAGE_LIST}" | tail -n ${IMAGES_PER_FILE} > "${TEXT_FILE}"
  fi
done

# Run the process on the GPUs, allowing multiple jobs per GPU
for ((i=0; i<TOTAL_JOBS; i++)); do
  GPU_ID=$((i % TOTAL_GPUS))
  CUDA_VISIBLE_DEVICES=${VALID_GPU_IDS[GPU_ID]} python ${RUN_FILE} \
    ${CONFIG_FILE} \
    ${CHECKPOINT} \
    --input "${INPUT}/image_paths_$((i+1)).txt" \
    --output-root="${OUTPUT}" ## add & to process in background
  # Allow a short delay between starting each job to reduce system load spikes
  sleep 1
done

# Wait for all background processes to finish
wait

# Remove the image list and temporary text files
rm "${IMAGE_LIST}"
for ((i=0; i<TOTAL_JOBS; i++)); do
  rm "${INPUT}/image_paths_$((i+1)).txt"
done

# Go back to the original script's directory
cd -

echo "Processing complete."
echo "Results saved to $OUTPUT"

```

sapiens_1b_seg_face-1024x768.py
```
# Copyright (c) Meta Platforms, Inc. and affiliates.
# All rights reserved.
#
# This source code is licensed under the license found in the
# LICENSE file in the root directory of this source tree.

_base_ = [
    "../../_base_/default_runtime.py",
]

##-----------------------------------------------------------------
# model_name = 'sapiens_0.6b'; embed_dim=1280; num_layers=32
model_name = "sapiens_1b"
embed_dim = 1536
num_layers = 40
# model_name = 'sapiens_2b'; embed_dim=1920; num_layers=48

##-----------------------------------------------------------------
image_size = (768, 1024)  ## width x height
data_preprocessor = dict(size=image_size)

patch_size = 16
num_epochs = 200

pretrained_checkpoint = (
    "/pathsapiens/seg/scripts/finetune/seg_face/pretrain/checkpoints/sapiens_1b/sapiens_1b_goliath_best_goliath_mIoU_7994_epoch_151.pth"
)

vis_every_iters = 100
# vis_every_iters=1

evaluate_every_n_epochs = 1

num_classes = 23 + 1  ## add background class

##--------------------------------------------------------------------------
# model settings
norm_cfg = dict(type="SyncBN", requires_grad=True)
data_preprocessor = dict(
    type="SegDataPreProcessor",
    mean=[123.675, 116.28, 103.53],
    std=[58.395, 57.12, 57.375],
    bgr_to_rgb=True,  ## convert from bgr to rgb for pretrained models
    pad_val=0,
    size=(image_size[1], image_size[0]),
    seg_pad_val=255,
)
model = dict(
    type="EncoderDecoder",
    data_preprocessor=data_preprocessor,
    pretrained=None,
    backbone=dict(
        type="mmpretrain.VisionTransformer",
        arch=model_name,
        img_size=(image_size[1], image_size[0]),
        patch_size=patch_size,
        qkv_bias=True,
        final_norm=True,
        drop_path_rate=0.0,
        with_cls_token=False,
        out_type="featmap",
        init_cfg=dict(type="Pretrained", checkpoint=pretrained_checkpoint),
    ),
    decode_head=dict(
        type="VitHead",
        in_channels=embed_dim,
        channels=768,
        deconv_out_channels=(
            768,
            768,
            768,
        ),  ## this will 2x at each step. so total is 8x
        deconv_kernel_sizes=(4, 4, 4),
        conv_out_channels=(768, 768, 768),
        conv_kernel_sizes=(1, 1, 1),
        num_classes=num_classes,
        norm_cfg=norm_cfg,
        align_corners=False,
        loss_decode=dict(type="CrossEntropyLoss", use_sigmoid=False, loss_weight=1.0),
    ),
    # model training and testing settings
    train_cfg=dict(),
    test_cfg=dict(mode="whole"),
)

##--------------------------------------------------------------------------
# optimizer
custom_imports = dict(
    imports=["mmseg.engine.optimizers.layer_decay_optim_wrapper"],
    allow_failed_imports=False,
)

## make sure the num_layers is same as the architecture
optim_wrapper = dict(
    optimizer=dict(type="AdamW", lr=5e-4, betas=(0.9, 0.999), weight_decay=0.1),
    paramwise_cfg=dict(
        num_layers=num_layers,
        layer_decay_rate=0.85,
        custom_keys={
            "bias": dict(decay_multi=0.0),
            "pos_embed": dict(decay_mult=0.0),
            "relative_position_bias_table": dict(decay_mult=0.0),
            "norm": dict(decay_mult=0.0),
        },
    ),
    constructor="LayerDecayOptimWrapperConstructor",
    clip_grad=dict(max_norm=1.0, norm_type=2),
)

param_scheduler = [
    dict(
        type="LinearLR", start_factor=1e-3, begin=0, end=400, by_epoch=False
    ),  # warm-up
    dict(
        type="PolyLR",
        eta_min=0.0,
        power=1.0,
        begin=0,
        end=num_epochs,
        by_epoch=True,
    ),
]

##--------------------------------------------------------------------------
train_cfg = dict(
    type="EpochBasedTrainLoop",
    max_epochs=num_epochs,
    val_interval=evaluate_every_n_epochs,
)
val_cfg = dict(type="ValLoop")
test_cfg = dict(type="TestLoop")
log_processor = dict(by_epoch=True)

# hooks
env_cfg = dict(
    mp_cfg=dict(mp_start_method="spawn", opencv_num_threads=0),
)

default_hooks = dict(
    timer=dict(type="IterTimerHook"),
    logger=dict(type="LoggerHook", interval=10),
    param_scheduler=dict(type="ParamSchedulerHook"),
    sampler_seed=dict(type="DistSamplerSeedHook"),
    checkpoint=dict(
        type="CheckpointHook", by_epoch=True, interval=1, max_keep_ckpts=-1
    ),
    visualization=dict(
        type="GeneralSegVisualizationHook",
        interval=vis_every_iters,
        max_samples=4,
        vis_image_width=384,
        vis_image_height=512,
    ),
)

##-----------------------------------------------------------------
# dataset settings
train_pipeline = [
    dict(type="LoadImage"),
    dict(type="LoadAnnotations"),
    dict(
        type="RandomResize",
        scale=(768, 1024),  ## width, height
        ratio_range=(0.2, 2.0),
        keep_ratio=True,
    ),
    dict(
        type="RandomCrop", crop_size=(1024, 768), cat_max_ratio=0.75
    ),  ## height, width
    dict(
        type="Resize", scale=(768, 1024), keep_ratio=False
    ),  ## in case if image was too small and random crop returned the original image
    dict(
        type="RandomRotate", prob=0.5, degree=45, seg_pad_val=0
    ),  ## the black pixels are set as background
    dict(
        type="RandomFlip", prob=0.5, swap_seg_labels=[(14, 23), (19, 20), (21, 22)]
    ),  ## for the new 28 classes,
    dict(type="PhotoMetricDistortion"),
    dict(type="PackSegInputs"),
]

test_pipeline = [
    dict(type="LoadImage"),
    dict(
        type="Resize", scale=(768, 1024), keep_ratio=False
    ),  ## this is width x height, 768 x 1024
    # add loading annotation after ``Resize`` because ground truth
    # does not need to do resize data transform
    dict(type="LoadAnnotations"),
    dict(type="PackSegInputs"),
]

##------------------------------------------------------------------------
dataset_train = dict(
    type="FaceDataset",
    data_root="/path/FaceSynthetics/dataset_100",
    reduce_zero_label=False,  # Whether to mark label zero as ignored.
    serialize_data=False,
)

train_datasets = [dataset_train]

train_dataloader = dict(
    batch_size=1,
    num_workers=4,
    persistent_workers=True,
    sampler=dict(type="DefaultSampler", shuffle=True),
    dataset=dict(
        type="CombinedDataset",
        metainfo=None,
        datasets=train_datasets,
        pipeline=train_pipeline,
    ),
)

##-----------------------------------------------------------------
val_evaluator = None
test_evaluator = None
val_dataloader = None
test_dataloader = None
val_cfg = None
test_cfg = None

```

**Comment by rawalkhirodkar** [comment]
@ShulingTang yes.

Please change the $DATASET variable from goliath to seg_face.

It should be the same config file you used for training.

**Comment by rawalkhirodkar** [comment]
@ShulingTang Also, if you are using your own segmentation classes, you need to provide the correct seg flip labels.

In your config file, modify train_pipeline.RandomFlip.seg_swap_labels accordingly.

**Comment by ShulingTang** [comment]
> @ShulingTang yes.
> 
> Please change the $DATASET variable from goliath to seg_face.
> 
> It should be the same config file you used for training.

Is work for me, thanks
